torch/csrc/dynamo/eval_frame.c

```diff
@@ -304,10 +305,100 @@ typedef struct cache_entry {
   struct cache_entry* next;
 } CacheEntry;
 
+#define Py_NONE_CACHE_ENTRY (CacheEntry*)Py_None
+
+static void destroy_cache_entry(CacheEntry* e);
+
+#define DECLARE_CACHE_ENTRY_ATTR(name) \
+static PyObject* CacheEntry_##name(CacheEntry* self, PyObject* _noargs) { \
+  PyObject* res = (PyObject*)self->name; \
+  Py_INCREF(res); \
+  return res; \
+}
+
+DECLARE_CACHE_ENTRY_ATTR(check_fn)
+DECLARE_CACHE_ENTRY_ATTR(code)
+DECLARE_CACHE_ENTRY_ATTR(next)
+
+static struct PyGetSetDef CacheEntry_properties[] = {
+    {"check_fn", (getter)CacheEntry_check_fn, NULL, NULL, NULL},
+    {"code", (getter)CacheEntry_code, NULL, NULL, NULL},
+    {"next", (getter)CacheEntry_next, NULL, NULL, NULL},
+    {NULL}};
+
+
+static PyObject* new_cache_entry(PyTypeObject* type, PyObject* args, PyObject* kwargs) {
+  CacheEntry *self;
+  self = (CacheEntry*) type->tp_alloc(type, 0);
+  if (self != NULL) {
+    // The corresponding decrefs for Py_None are in init_cache_entry.
+    Py_INCREF(Py_None);
+    self->check_fn = Py_None;
+    Py_INCREF(Py_None);
+    self->code = (PyCodeObject*)Py_None;
+    Py_INCREF(Py_None);
+    self->next = Py_NONE_CACHE_ENTRY;
+  }
+  return (PyObject*)self;
+}
+
+
+static int init_cache_entry(CacheEntry* self, PyObject* args, PyObject* kwds) {
+  PyObject* check_fn = NULL;
+  PyCodeObject* code = NULL;
+  CacheEntry* next = NULL;
+
+  static char *kwlist[] = {"check_fn", "code", "next", NULL};
+
+  int ret = PyArg_ParseTupleAndKeywords(
+    args, kwds, "OOO", kwlist,
+    &check_fn, &code, &next);
+
+  if (!ret) return -1;
+
+  if (check_fn) {
+    PyObject* tmp = self->check_fn;
+    self->check_fn = check_fn;
```

Comment:
The parser here doesn't give you a new reference, it's a borrow. So if you want to stash it in self I think you have to `Py_INCREF` it.

Reply:
So, I am confused here. The `check_fn` is originating from the `convert_frame.py`, and when `call_callback` returns the GuardedCode object to `eval_frame.c`, CacheEntry steals the reference. This particular line in the init function is just an intermediate stage in the plumbing, with finally extra scratch scape becoming the owner of the reference. Therefore, I did not INCREF.

Also, a few test cases fail if I do INCREF here. And I came up with above reasoning to satisfy myself. Does it make sense?

Reply:
See https://github.com/pytorch/pytorch/pull/107405#discussion_r1299505687